--- a/CMakeLists.txt	2023-05-21 09:54:07.000000000 +0300
+++ b/CMakeLists.txt	2024-12-10 02:54:24.488887737 +0300
@@ -1617,7 +1617,7 @@
 #
 
 set(EVENT_INSTALL_CMAKE_DIR
-    "${CMAKE_INSTALL_LIBDIR}/cmake/libevent")
+    "lib/cmake/libevent")
 
 export(PACKAGE libevent)
 
--- a/cmake/AddEventLibrary.cmake	2023-05-21 09:54:07.000000000 +0300
+++ b/cmake/AddEventLibrary.cmake	2024-12-10 02:59:09.980439955 +0300
@@ -27,7 +27,7 @@
     configure_file("lib${LIB_NAME}.pc.in" "lib${LIB_NAME}.pc" @ONLY)
     install(
         FILES "${CMAKE_CURRENT_BINARY_DIR}/lib${LIB_NAME}.pc"
-        DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig"
+        DESTINATION "lib/pkgconfig"
     )
 endmacro()
 
@@ -140,7 +140,7 @@
             set_target_properties(
                 "${LIB_NAME}_shared" PROPERTIES
                 OUTPUT_NAME "${LIB_NAME}-${EVENT_PACKAGE_RELEASE}.${CURRENT_MINUS_AGE}"
-                INSTALL_NAME_DIR "${CMAKE_INSTALL_LIBDIR}"
+                INSTALL_NAME_DIR "lib"
                 LINK_FLAGS "-compatibility_version ${COMPATIBILITY_VERSION} -current_version ${COMPATIBILITY_VERSION}.${EVENT_ABI_LIBVERSION_REVISION}")
         else()
             math(EXPR CURRENT_MINUS_AGE "${EVENT_ABI_LIBVERSION_CURRENT}-${EVENT_ABI_LIBVERSION_AGE}")
